be/src/exec/vectorized/chunk_sorter_heapsorter.h

```diff
@@ -0,0 +1,263 @@
+#include "runtime/primitive_type.h"
+
+namespace starrocks::vectorized {
+
+namespace detail {
+using DataSegmentPtr = std::shared_ptr<DataSegment>;
+// this class will be delete by self
+// so we made desctuctor private to avoid alloc memory in stack
+class ChunkHolder {
+public:
+    ChunkHolder(DataSegmentPtr&& ref_value) : _ref_count(0), _ref_val(std::move(ref_value)) {}
+    ChunkHolder(const ChunkHolder&) = delete;
+    void unref() noexcept {
+        DCHECK_GT(_ref_count, 0);
+        _ref_count--;
+        if (_ref_count == 0) {
+            delete this;
+        }
+    }
+    void ref() noexcept { _ref_count++; }
+
+    int ref_count() const { return _ref_count; }
+
+    DataSegmentPtr& value() { return _ref_val; }
+
+private:
+    ~ChunkHolder() noexcept = default;
+    int _ref_count;
+    DataSegmentPtr _ref_val;
+};
+
+struct ChunkRowCursor {
+public:
+    ChunkRowCursor(int row_id, ChunkHolder* holder) : _row_id(row_id), _holder(holder) { _holder->ref(); }
+    ChunkRowCursor(const ChunkRowCursor& other) {
+        _row_id = other._row_id;
+        _holder = other._holder;
+        _holder->ref();
+    }
+
+    ChunkRowCursor& operator=(const ChunkRowCursor& other) {
+        if (_holder) {
+            _holder->unref();
+        }
+        _row_id = other._row_id;
+        _holder = other._holder;
+        _holder->ref();
+        return *this;
+    }
+
+    ChunkRowCursor(ChunkRowCursor&& other) {
+        _row_id = other._row_id;
+        _holder = other._holder;
+        other._holder = nullptr;
+    }
+
+    ChunkRowCursor& operator=(ChunkRowCursor&& other) {
+        std::swap(_row_id, other._row_id);
+        std::swap(_holder, other._holder);
+        return *this;
+    }
+
+    ~ChunkRowCursor() noexcept {
+        if (_holder) {
+            _holder->unref();
+        }
+    }
+    int row_id() const { return _row_id; }
+
+    int ref_count() const { return _holder->ref_count(); }
+
+    const DataSegmentPtr& data_segment() const { return _holder->value(); }
+
+private:
+    size_t _row_id;
+    ChunkHolder* _holder;
+};
+
+template <typename T, typename Sequence, typename Compare>
+class SortingHeap {
+public:
+    SortingHeap(Compare comp) : _comp(std::move(comp)) {}
+
+    bool is_valid() const { return !_queue.empty(); }
+
+    const T& top() { return _queue.front(); }
+
+    size_t size() { return _queue.size(); }
+
+    bool empty() { return _queue.empty(); }
+
+    T& next_child() { return _queue[_next_child_index()]; }
+
+    void reserve(size_t reserve_sz) { _queue.reserve(reserve_sz); }
+
+    void replace_top(T&& new_top) {
+        *_queue.begin() = std::move(new_top);
+        update_top();
+    }
+
+    void remove_top() {
+        std::pop_heap(_queue.begin(), _queue.end(), _comp);
+        _queue.pop_back();
+    }
+
+    void push(T&& rowcur) {
+        _queue.emplace_back(std::move(rowcur));
+        std::push_heap(_queue.begin(), _queue.end(), _comp);
+    }
+
+    Sequence&& sorted_seq() {
```

Comment:
Why use the `Sequence` name here?

Reply:
CXX STL priority_queue use Sequenece as Container name.I refer to the naming convention of STL